### PR TITLE
ISPN-11366 Remove Deprecated ConsistentHash methods

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-11366

### CODE REVIEW

**Comment by ryanemerson** [comment]
@danberindei I'm having issues removing `ConsistentHash::getNumOwners` still, hence why I have omitted it in this PR. The changes are causing the rebalance to timeout in a lot of the tests, but I can't figure out what's wrong. I'm also not that confident that my changes in the latest commit are right. This is my wip branch: https://github.com/ryanemerson/infinispan/tree/ISPN-11366-getNumOwners.

I don't really want to spend too much more time on this as I am making the changes mainly to remove the need to serialize the`Hash` implementation in the `ConsistentHash` to enable the GlobalMarshaller changes.

File: core/src/main/java/org/infinispan/distribution/group/impl/PartitionerConsistentHash.java [hunk 1]
```diff
@@ -41,11 +41,6 @@ public int getNumOwners() {
       return ch.getMembers();
    }
 
-   @Override
-   public int getSegment(Object key) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/distribution/group/impl/PartitionerConsistentHash.java]: We should be able to remove `PartitionerConsistentHash` altogether.

File: core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java [hunk 2]
```diff
@@ -483,29 +483,30 @@ public Object visitWriteOnlyKeyCommand(InvocationContext ctx, WriteOnlyKeyComman
...
-                     .invokeCommand(backupOwner, copy, SingleResponseCollector.validOnly(),
-                                    rpcManager.getSyncRpcOptions())
-                     .toCompletableFuture();
-            } else {
-               rpcManager.sendTo(backupOwner, copy, DeliverOrder.PER_SENDER);
-            }
+      LocalizedCacheTopology topology = checkTopologyId(command);
+      ConsistentHash ch = topology.getWriteConsistentHash();
+      // We have already checked that the command topology is actual, so we can assume that we really are primary owner
+      Map<Address, IntSet> backups = backupOwnersOfSegments(ch, IntSets.from(ch.getPrimarySegmentsForOwner(rpcManager.getAddress())));
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java]: Could use `LocalizedCacheTopology#getLocalReadSegments`

File: core/src/test/java/org/infinispan/distribution/topologyaware/TopologyAwareConsistentHashFactoryTest.java [hunk 3]
```diff
@@ -62,45 +65,50 @@ public void testNumberOfOwners() {
...
       updateConsistentHash(2);
-      assertEquals(ch.locateOwners(testAddresses[0]).size(), 1);
+      assertNumberOwners(testAddresses[0], 1);
 
       addNode(testAddresses[1], "m1", null, null);
 
       updateConsistentHash(1);
-      for (Address testAddress : testAddresses) {
-         assertEquals(ch.locateOwners(testAddress).size(), 1);
+      for (Address testAddress : ch.getMembers()) {
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/distribution/topologyaware/TopologyAwareConsistentHashFactoryTest.java]: This is kind of outdated, the test should iterate over the segments and check the owners of each segment instead.

File: core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java [hunk 4]
```diff
@@ -148,7 +148,7 @@ public boolean isSatisfied() throws Exception {
       ConsistentHash ch = cache(a0).getAdvancedCache().getDistributionManager().getReadConsistentHash();
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java]:
```suggestion
      LocalizedCacheTopology cacheTopology = cache(a0).getAdvancedCache().getDistributionManager().getCacheTopology();
```

File: core/src/test/java/org/infinispan/statetransfer/TestKey.java [hunk 5]
```diff
@@ -32,10 +34,11 @@ public TestKey(String name, int segmentId, ConsistentHash ch) {
       this.name = name;
 
       Random rnd = new Random();
+      KeyPartitioner keyPartitioner = new HashFunctionPartitioner(ch.getNumSegments());
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/statetransfer/TestKey.java]: I think the parameter should be the KP, not the CH.

**Comment by ryanemerson** [comment]
@danberindei Should be good now

**Comment by danberindei** [comment]
Conflicts @ryanemerson

**Comment by ryanemerson** [comment]
Should be good now

File: core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java [hunk 6]
```diff
@@ -124,7 +124,8 @@ public void setMembersToUse(List<Address> membersToUse) {
    @SerializeWith(Externalizer.class)
    public static class Default extends ControlledConsistentHashFactory<DefaultConsistentHash> {
       public Default(int primaryOwnerIndex, int... backupOwnerIndexes) {
-         super(new DefaultTrait(), primaryOwnerIndex, backupOwnerIndexes);
+         super(new DefaultTrait(),
+               primaryOwnerIndex, backupOwnerIndexes);
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java]: Can stay on the same line

File: core/src/test/java/org/infinispan/util/BaseControlledConsistentHashFactory.java [hunk 7]
```diff
@@ -99,23 +98,32 @@ public CH union(CH ch1, CH ch2) {
...
       CH union(CH ch1, CH ch2);
 
       boolean requiresPrimaryOwner();
+
+      int getNumOwners(CH ch);
    }
 
    public static class DefaultTrait implements Trait<DefaultConsistentHash> {
+
+      public DefaultTrait() {
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/util/BaseControlledConsistentHashFactory.java]: No longer needed

**Comment by danberindei** [comment]
Merged, thanks Ryan!

**Pull request merged**